File: LeetCode/[M] - 875. Koko Eating Bananas/solution.go
package main

import "slices"

func minEatingSpeed(piles []int, h int) int {
	l, r := 1, slices.Max(piles)
	k := r
	for l <= r {
		kNew := l + (r-l)/2
		if isValid(kNew, piles, h) {
			k = kNew
			r = kNew - 1
		} else {
			l = kNew + 1
		}
	}
	return k
}

func isValid(kNew int, piles []int, h int) bool {
	hNew := 0
	for _, pile := range piles {
		if pile%kNew == 0 {
			hNew += pile / kNew
		} else {
			hNew += pile/kNew + 1
		}
	}
	return hNew <= h
}

File: LeetCode/[M] - 875. Koko Eating Bananas/solution.py
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)
        k = right
        while left <= right:
            kNew = left + (right - left) // 2
            if self.isValid(kNew, piles, h):
                k = kNew
                right = kNew - 1
            else:
                left = kNew + 1
        return k

    def isValid(self, kNew, piles, h):
        hNew = 0
        for pile in piles:
            hNew += (pile // kNew) + (pile % kNew != 0)
        return hNew <= h
